src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/PipelineExtensions.cs

```diff
@@ -68,7 +70,9 @@ public static ValueTask<ArraySegment<byte>> PeekAsync(this IPipeReader pipelineR
                 }
                 finally
                 {
-                    pipelineReader.Advance(result.Buffer.Start, result.Buffer.Start);
+                    pipelineReader.Advance(result.Buffer.Start, result.Buffer.IsEmpty
+                        ? result.Buffer.End
+                        : result.Buffer.Start);
                 }
 
                 readingTask = pipelineReader.ReadAsync();
```

Comment:
I think I misunderstood you on my first read.

But `ConsumedBytes` calls `ReadAsync` as well, so this call has to call `Advance` before returning otherwise that `ReadAsync` will fail.